comparison/comparison.go

```diff
@@ -0,0 +1,509 @@
+		cancelCh:    make(chan struct{}),
+	}
+
+	var i int
+	deployments := map[string]string{}
+	for _, lt := range cfg.LoadTests {
+		key := string(lt.Type) + string(lt.DBEngine)
+		var dp *deploymentInfo
+		if id, ok := deployments[key]; !ok {
+			dp = &deploymentInfo{config: *deployerCfg}
+			dp.config.DBInstanceEngine = "aurora-" + string(lt.DBEngine)
+			dp.config.ClusterName = fmt.Sprintf("%s%d", dp.config.ClusterName, i)
+			id = fmt.Sprintf("deployment%d", i)
+			cmp.deployments[id] = dp
+			deployments[key] = id
+			i++
+		} else {
+			dp = cmp.deployments[id]
+		}
+		dp.loadTests = append(dp.loadTests, lt)
+	}
+
+	return cmp, nil
+}
+
+func (c *Comparison) deploymentAction(action func(t *terraform.Terraform) error) error {
+	var wg sync.WaitGroup
+	wg.Add(len(c.deployments))
+	errsCh := make(chan error, len(c.deployments))
+	for id, dp := range c.deployments {
+		go func(id string, dp *deploymentInfo) {
+			defer wg.Done()
+			t := terraform.New(id, &dp.config)
+			defer t.Cleanup()
+			if err := action(t); err != nil {
+				errsCh <- fmt.Errorf("deployment action failed: %w", err)
+			}
+		}(id, dp)
+	}
+	wg.Wait()
+	close(errsCh)
+	var err error
+	for err = range errsCh {
+		mlog.Error(err.Error())
+	}
+	return err
+}
+
+func (c *Comparison) getLoadTestsCount() int {
+	var count int
+	for _, dp := range c.deployments {
+		for range dp.loadTests {
+			count++
+		}
+	}
+	return count
+}
+
+func runBoundedLoadTest(t *terraform.Terraform, coordConfig *coordinator.Config, d time.Duration, cancelCh <-chan struct{}) (coordinator.Status, error) {
+	var err error
+	var status coordinator.Status
+	mlog.Info("starting bounded load-test")
+	if err := t.StartCoordinator(coordConfig); err != nil {
+		return status, err
+	}
+
+	var canceled bool
+	select {
+	case <-cancelCh:
+		mlog.Info("cancelling load-test")
+		canceled = true
+	case <-time.After(d):
+	}
+
+	mlog.Info("stopping bounded load-test")
+	status, err = t.StopCoordinator()
+	if err != nil {
+		return status, err
+	}
+
+	if canceled {
+		return status, errors.New("canceled")
+	}
+
+	// TODO: remove this once MM-30326 has been merged and a new release
+	// published.
+	status.StopTime = time.Now()
+
+	return status, nil
+}
+
+func runUnboundedLoadTest(t *terraform.Terraform, coordConfig *coordinator.Config, cancelCh <-chan struct{}) (coordinator.Status, error) {
+	mlog.Info("starting unbounded load-test")
+	if err := t.StartCoordinator(coordConfig); err != nil {
+		return coordinator.Status{}, err
+	}
+
+	for {
+		status, err := t.GetCoordinatorStatus()
+		if err != nil {
+			return status, err
+		}
+
+		if status.State == coordinator.Done {
+			mlog.Info("unbounded load-test has completed")
+			return status, nil
+		}
+
+		if status.State != coordinator.Running {
+			return status, errors.New("coordinator should be running")
+		}
+
+		select {
+		case <-cancelCh:
+			mlog.Info("cancelling load-test")
+			if status, err := t.StopCoordinator(); err != nil {
```

Comment:
It seems like we are not stopping the coordinator if we get an error from `GetCoordinatorStatus`. Should we place this inside a defer block and do appropriate checks there so that the coordinator is always stopped?

Reply:
Agree, a deferred cleanup function makes total sense in this case.